### ISSUE TITLE
[Tech Design] Terra Select Responsive Dropdown

### ISSUE THREAD

**Issue by dev-cprice** [cerner/terra-core#2405]
# Tech Design

## Description

Tech design for https://github.com/cerner/terra-core/issues/2383

### TLDR

* Consolidate use of [Dropdown](https://github.com/cerner/terra-core/blob/master/packages/terra-form-select/src/_Dropdown.jsx) and [DropdownMenu](https://github.com/cerner/terra-core/blob/master/packages/terra-form-select/src/_Menu.jsx)
  * `<DropdownMenu />` is an internal implementation detail of `<Dropdown />` and it's unnecessarily complicating the component by rendering the menu as a renderProp in the `<Select />` component
* Implement (name pending) `<ResponsiveDropdown />` (or `<Modal />`, whatever the UX defined "responsive dropdown" is)
* Remove render props pattern in [Select.jsx](https://github.com/cerner/terra-core/blob/master/packages/terra-form-select/src/Select.jsx#L238-L243)
  * In [Frame](https://github.com/cerner/terra-core/blob/master/packages/terra-form-select/src/_Frame.jsx#L728-L754), if `this.state.isOpen`, then use either the `<Dropdown />` or `<Modal />`
* Flex between `<Dropdown />` and `<Modal />` based on if we're mobile or at `'small'` or `'tiny'` breakpoints
  * Add `isMobile` method to [SharedUtil](https://github.com/cerner/terra-core/blob/master/packages/terra-form-select/src/_SharedUtil.js) which uses the `UserAgent` to check for `iPhone`, `iPad`, and `Android`
  * Use [ActiveBreakpointContext](https://github.com/cerner/terra-core/blob/master/packages/terra-breakpoints/src/ActiveBreakpointContext.jsx) to determine the `activeBreakpoint`
  * `const useModal = isMobile() || activeBreakpoint === 'tiny' || activeBreakpoint === 'small';`

#### Dropdown, DropdownMenu, and the Unnecessary `renderProps` Pattern

`<DropdownMenu />` is just an internal implementation detail of `<Dropdown />`. We don't need to be rendering it via render props in `<Select />`. We can just render the `<Frame />` and delegate the dropdown and responsive behavior to the Frame. This allows us to reduce the complexity of the `Select` component while also giving us the ability to add new behavior easily to the `Frame`.

#### Responsive Dropdown Implementation

***NOTE:*** See additional screenshots below for visuals.

The responsive dropdown workflow will live inside a `DialogModal`. When rendered, it will have an action header and action footer, while the content will be displayed with an input styled just like the current dropdown variant (including tags/pills) and a list of all options.

##### Action Header

We will need to give the action header a label to display as the title. The action header will call the responsive dropdown component's `onClose` prop when the action header's close button is pressed/clicked and close the modal.

##### Content

The content will contain an input (for multiple/search/tag variants) and a list of options as well as any extra visual indicators for workflows that exist outside the happy-path (e.g. "No results found") or normal usage (e.g. "Enter ‘x’ characters to search").

##### Action Footer

The action footer will have two buttons. The start button is labeled `"Clear All"` and will call the responsive dropdown's `onClear` prop when pressed/clicked. The end button is labeled `"Apply"` and will simply _approve/commit_ the state change authored by the user. It will call the responsive dropdown's `onApply` prop when pressed/clicked. It is assumed that changes will only be committed when the `Apply` button is used, while also closing the modal.

#### isMobile()

```js
class SharedUtil {
  // ...

  static isMobile() {
    const { userAgent } = navigator;
    const isAndroid = userAgent.includes('Android');
    const isIOS = userAgent.includes('iPhone') || userAgent.includes('iPad');

    return isAndroid || isIOS;
  }
}
```

#### Flexing between <Dropdown /> and <Modal /> workflows

Right now in [Frame](https://github.com/cerner/terra-core/blob/master/packages/terra-form-select/src/_Frame.jsx#L728-L754), we render a `<Dropdown />` (using the renderProps pattern to pass props to a `<DropdownMenu />` in `<Select />`) based on the boolean `this.state.isOpen`. We can use this plus another boolean to flex between `<Dropdown />` and `<Modal />` workflows. Using the above `isMobile()` method and the [ActiveBreakpointContext](https://github.com/cerner/terra-core/blob/master/packages/terra-breakpoints/src/ActiveBreakpointContext.jsx), we can determine if we want to render the `<Modal />` or not:

```jsx
const activeBreakpoint = useContext(ActiveBreakpointContext);
const useModal = isMobile() || activeBreakpoint === 'tiny' || activeBreakpoint === 'small';
const { isOpen } = this.state;

return (
  <div {...stuff}>
    {isOpen &&
      useModal ? <Modal {...modalProps} /> : <Dropdown {...dropdownProps} />
    }
  </div>
);
```

## Additional Context / Screenshots

![image](https://user-images.githubusercontent.com/34115417/57952853-0b449080-78b4-11e9-89df-feb6443755a5.png)

## @ Mentions

@bjankord
@StephenEsser

**Comment by dev-cprice** [comment]
Also of note: rudimentary tests with a similar layout as above show iOS voiceover working as expected since all the children of the dialog modal are siblings of the input, rather than visually appearing to be.

**Comment by bjankord** [comment]
+1 on tech design.

**Comment by StephenEsser** [comment]
A little more context about the render prop so the decision to remove it isn't done under misapprehension.

The Select menu was originally designed to allow a custom implementation of the menu within the dropdown. The render prop was added to allow an entry point for this functionality. A user could implement a Select dropdown using the Frame and provide their own menu which would receive all of the require props from the render prop. 

```jsx
<Frame dropdown={dropdownProps => ( 
  <CustomMenu {...dropdownProps} />)} 
/>
```

This would allow a user to create a custom menu implementation to meet application specific requirements. This requirement is also why the Frame itself exists as a separate component from the Select. 

With the introduction of the accessibility changes, general complexity of this component, and the drive towards a standard behavior a custom menu is becoming increasingly more difficult and discouraged to implement.

If we decide to remove the custom render functionality we should also investigate merging the Frame into the Select to cut down on complexity and potential extra renders.

**Comment by dev-cprice** [comment]
@StephenEsser good point!

when writing up the tech design, I had a similar thought: _if we remove the renderProp pattern, we're basically just rendering the frame directly now so why do we need it?_

👍 👍 👍

**Comment by StephenEsser** [comment]
Do you foresee any issues customizing the Dialog Modal to display with a different positioning and height? The height would likely need to flex based on the Select variant that will invoke the keyboard. 

The same question would apply to the Action Header and Action Footer. Are we going to alter the existing components to match the callouts or will they display with the default style?

- Background Colors
- Close Icon
- Apply and Clear All Buttons
- Borders

Will `Clear All` within the modal be conditionally rendered based on the presence of `allowClear`?

What are the disadvantages / advantages to using the components over creating a simple flex container?

Is the Input "*sticky*"? Does it scroll with the set of available options or does it remain fixed at the top of the Modal? The callouts look like it might scroll with the options, but we should get some clarification from UX on this.

( This is if it remains fixed at the top of the Modal )

What happens when the selected options width grows to be the height of the Dialog Modal? Previously this problem was somewhat mitigated by the ability to scroll the page. With the introduction of the Dialog Modal the page is not scrollable. When enough options have been selected the Options will no longer be visible. 

We discussed offline about making the label of the Select required to enforce a label to be present when presented in a dialog.

**Comment by dkasper-was-taken** [comment]
We'll want to avoid using the user agent sniffing. There is a significant variability to the adherence to APIs and standards. As well as is against terra's general guideline to avoid device specific implementations, and code in favor of viewports and/or parent element sizing exclusively.

Here's also an article about it on MDN, about it shouldn't be used in this fashion:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Browser_detection_using_the_user_agent

We'll likely want to avoid the terra-dialog-modal, and use the terra-abstract-modal with specific implementation and styling.

**Comment by dev-cprice** [comment]
@StephenEsser 

>The same question would apply to the Action Header and Action Footer. Are we going to alter the existing components to match the callouts or will they display with the default style?

>What are the disadvantages / advantages to using the components ( Action Header / Footer ) over creating a simple flex container?

We can definitely roll our own headers & footers here. Doing so will help decouple the implementations between these components and allow for more control over styling (this appears to be the standard practice in terra).

>Will Clear All within the modal be conditionally rendered based on the presence of allowClear?

That seems reasonable.

> Is the Input "sticky"? Does it scroll with the set of available options or does it remain fixed at the top of the Modal? The callouts look like it might scroll with the options, but we should get some clarification from UX on this.

Yeah, I'll reach out to UX on this. From this image: 
![image](https://user-images.githubusercontent.com/34115417/58276099-328bd980-7d5c-11e9-91a4-f8d4128f0e71.png)

it does look like the input + list items scroll together.

>What happens when the selected options height grows to be the height of the Dialog Modal? Previously this problem was somewhat mitigated by the ability to scroll the page. With the introduction of the Dialog Modal the page is not scrollable. When enough options have been selected the Options will no longer be visible.

From Dave's comment, I'll be using the abstract modal now. I wasn't aware of the dialog modal (or abstract modal) not allowing scrolling. Looking now, I see that `overflow: hidden;` is set for the [abstract modal](https://github.com/cerner/terra-framework/blob/master/packages/terra-abstract-modal/src/AbstractModal.module.scss#L8), but the dialog modal's body is set to [overflow-y: auto;](https://github.com/cerner/terra-framework/blob/master/packages/terra-dialog-modal/src/DialogModal.module.scss#L82)

>We discussed offline about making the label of the Select required to enforce a label to be present when presented in a dialog.

This seems reasonable (or making the fallback what it currently is). Since this will be a breaking change, it makes sense to add props and/or change required-ness.

---

@dkasper-was-taken 

> We'll want to avoid using the user agent sniffing.

We already do it in the select shared utils [here](https://github.com/cerner/terra-core/blob/master/packages/terra-form-select/src/_SharedUtil.js#L6-L12). Are you suggesting removing that? One of the use cases that @bjankord brought up is that we'll want iPads to also use the responsive dropdown since they'll also succumb to the iOS voiceover a11y concerns driving this redesign. As far as I'm aware, the iPad breakpoints can be pretty big and may map to large (or even huge).

Here's what the simulators for an ipad pro show when visiting https://engineering.cerner.com/terra-ui/#/components/terra-breakpoints/breakpoints/about

<img width="993" alt="Screen Shot 2019-05-23 at 1 28 33 PM" src="https://user-images.githubusercontent.com/34115417/58277154-b2b33e80-7d5e-11e9-9fc7-3e54fec1487b.png">
<img width="1259" alt="Screen Shot 2019-05-23 at 1 28 20 PM" src="https://user-images.githubusercontent.com/34115417/58277156-b47d0200-7d5e-11e9-940d-9c67a5e35a20.png">




> We'll likely want to avoid the terra-dialog-modal, and use the terra-abstract-modal with specific implementation and styling.

👍

**Comment by emilyrohrbough** [comment]
Just a few questions from briefly looking at this:

- My initial thought was how do you close the modal? It appears there is an `x` in the input, but I would assumed that is to clear all the content. However, there is an `x` on each option and a `clear all` option in the footer. Is clicking `apply` the way to close?

- Is there a max height on the input that displays the selected options? Should there be an input? How will this height and the dropdown height be managed when keyboard is rendered?

- For determining the mobile sizing, should we match what is currently implemented in terra-time-input? https://github.com/cerner/terra-framework/blob/master/packages/terra-time-input/src/TimeUtil.js#L191 
 I agree with Dave that determining the agent is likely not a great approach in supporting multiple browser and browser versions. Could there be screen reader issues on non-IOS tablets that use firefox or

**Comment by bjankord** [comment]
Originally I was thinking we'd need to use UA detection to ensure this modal shows up on all iOS devices (including the larger iPad Pro devices), [even though I'm not a big fan of user agent sniffing and instead prefer feature detection](https://www.brettjankord.com/2013/01/22/beyond-device-detection/).

I'd recommend we check for the following conditions to trigger showing the modal:
* The large breakpoint is not active and it is not the default variant or the browser supports touch events and is not the default variant.

The first condition covers most "mobile" devices and we don't want to trigger the modal for the default variant, that can keep using the dropdown
The second condition covers the larger iPad devices (we need to include these since the issue we are working to fix with with is with VoiceOver on iOS devices). For the second condition I'd only check for `'ontouchstart' in window` and not worry about max-touch-points.

Let's move forward with a PR for this @dev-cprice and we can iterate on it on the PR and address questions as we get this built out some more.

**Comment by bjankord** [comment]
Update: We've talked about this offline and have landed on implementing a solution that renders the input within the portal instead of rendering the select as a modal on mobile devices. Closing this issue as we've decided on a direction to go with this that won't require the implementation described in the tech design.
